From 5c03a5dcec18916f97422289ecc5d628ff50239d Mon Sep 17 00:00:00 2001
From: jjacky <i.am.jack.mail@gmail.com>
Date: Wed, 31 Jul 2013 13:56:08 +0200
Subject: [PATCH 2/2] treeview: Add gtk_tree_view_start_rubber_banding()

This allows to start a rubber banding operation based on one's own criteria,
e.g. only from blank space, a specific column/button, etc

Signed-off-by: jjacky <i.am.jack.mail@gmail.com>
---
 gtk/gtk.symbols   |  1 +
 gtk/gtktreeview.c | 51 +++++++++++++++++++++++++++++++++++++++++++++++++++
 gtk/gtktreeview.h |  2 ++
 3 files changed, 54 insertions(+)

diff --git a/gtk/gtk.symbols b/gtk/gtk.symbols
index 030a06c..2acb4b8 100644
--- a/gtk/gtk.symbols
+++ b/gtk/gtk.symbols
@@ -3618,6 +3618,7 @@ gtk_tree_view_set_tooltip_cell
 gtk_tree_view_set_tooltip_column
 gtk_tree_view_set_tooltip_row
 gtk_tree_view_set_vadjustment
+gtk_tree_view_start_rubber_banding
 gtk_tree_view_unset_rows_drag_dest
 gtk_tree_view_unset_rows_drag_source
 gtk_true
diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index 929e259..c92a130 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -15796,6 +15796,57 @@ gtk_tree_view_get_rubber_banding (GtkTreeView *tree_view)
 }
 
 /**
+ * gtk_tree_view_start_rubber_banding:
+ * @tree_view: a #GtkTreeView
+ * @event: a #GdkEventButton from the ::button-press-event triggering the rubber
+ * banding
+ *
+ * Initiates a new rubber banding operation. This will do nothing and return
+ * %FALSE if one is already in place, or the selection mode isn't
+ * %GTK_SELECTION_MULTIPLE.
+ *
+ * You don't need to set ::rubber-banding for this to work, as the property only
+ * sets whether rubber banding operations are automatically started by GTK or
+ * not. Obviously, you should only call this from ::button-press-event using the
+ * associated #GdkEvenButton for things to work correctly.
+ *
+ * Return: %TRUE if a new rubber banding was initiated in @tree_view
+ *
+ * Since: 3.10
+ */
+gboolean
+gtk_tree_view_start_rubber_banding (GtkTreeView     *tree_view,
+                                    GdkEventButton  *event)
+{
+  GdkModifierType modify_mod_mask;
+  GdkModifierType extend_mod_mask;
+
+  g_return_val_if_fail (GTK_IS_TREE_VIEW (tree_view), FALSE);
+  g_return_val_if_fail (event != NULL, FALSE);
+
+  if (tree_view->priv->rubber_band_status != RUBBER_BAND_OFF
+      || gtk_tree_selection_get_mode (tree_view->priv->selection) == GTK_SELECTION_MULTIPLE)
+    return FALSE;
+
+  extend_mod_mask = gtk_widget_get_modifier_mask ((GtkWidget *) tree_view,
+      GDK_MODIFIER_INTENT_EXTEND_SELECTION);
+  modify_mod_mask = gtk_widget_get_modifier_mask ((GtkWidget *) tree_view,
+      GDK_MODIFIER_INTENT_MODIFY_SELECTION);
+
+  tree_view->priv->press_start_x = event->x;
+  tree_view->priv->press_start_y = event->y + tree_view->priv->dy;
+  tree_view->priv->rubber_band_x = event->x;
+  tree_view->priv->rubber_band_y = event->y + tree_view->priv->dy;
+  tree_view->priv->rubber_band_status = RUBBER_BAND_MAYBE_START;
+  if ((event->state & modify_mod_mask) == modify_mod_mask)
+    tree_view->priv->rubber_band_modify = TRUE;
+  if ((event->state & extend_mod_mask) == extend_mod_mask)
+    tree_view->priv->rubber_band_extend = TRUE;
+
+  return TRUE;
+}
+
+/**
  * gtk_tree_view_is_rubber_banding_active:
  * @tree_view: a #GtkTreeView
  * 
diff --git a/gtk/gtktreeview.h b/gtk/gtktreeview.h
index cdd94f0..ab59cd2 100644
--- a/gtk/gtktreeview.h
+++ b/gtk/gtktreeview.h
@@ -458,6 +458,8 @@ gboolean gtk_tree_view_get_hover_expand      (GtkTreeView          *tree_view);
 void     gtk_tree_view_set_rubber_banding    (GtkTreeView          *tree_view,
 					      gboolean              enable);
 gboolean gtk_tree_view_get_rubber_banding    (GtkTreeView          *tree_view);
+gboolean gtk_tree_view_start_rubber_banding  (GtkTreeView          *tree_view,
+                                              GdkEventButton       *event);
 
 gboolean gtk_tree_view_is_rubber_banding_active (GtkTreeView       *tree_view);
 
-- 
1.8.3.4

